prompb/io/prometheus/write/v2/symbols.go

```diff
@@ -73,11 +77,16 @@ func (t *SymbolsTable) Reset() {
 }
 
 // desymbolizeLabels decodes label references, with given symbols to labels.
-func desymbolizeLabels(b *labels.ScratchBuilder, labelRefs []uint32, symbols []string) labels.Labels {
+// This function requires labelRefs to have an even number of elements (name-value pairs). It will return an error if this invariant is violated.
+func desymbolizeLabels(b *labels.ScratchBuilder, labelRefs []uint32, symbols []string) (labels.Labels, error) {
+	if len(labelRefs)%2 != 0 {
+		return labels.EmptyLabels(), fmt.Errorf("invalid labelRefs length %d", len(labelRefs))
+	}
+
 	b.Reset()
 	for i := 0; i < len(labelRefs); i += 2 {
 		b.Add(symbols[labelRefs[i]], symbols[labelRefs[i+1]])
```

Comment:
We could also check if `len(symbols) > labelRefs[i] && len(symbols) > labelRefs[i
+1] ` if we are already at this (another potential source of panics). WDYT?

```suggestion
        nameRef, valueRef :=   labelRefs[i], labelRefs[i+1] 
        if len(symbols) <= nameRef || len(symbols) <= valueRef {
          return labels.EmptyLabels(), fmt.Errorf("labelRefs [%d=%d] outside of symbols table %d", nameRef, valueRef, len(symbols))
        }
		b.Add(symbols[nameRef], symbols[valueRef])
```

Reply:
IF you agree, let's also add tests for that.

Reply:
Wouldn't it have to be:
``` 
if len(symbols) < i || len(symbols) < i+1 {
    // error case
}
b.Add(symbols[labelRefs[i]], symbols[labelRefs[i+1]])
```

As soon as we try and access the items at indices i and i+1 we have the potential panic. I'd also be curious to see the changes (if any) in a `BenchmarkDesymbolizeLabels` for the happy path of 'entire symbol table is valid'.

Reply:
Hi @bwplotka, I have added the suggested changes at https://github.com/prometheus/prometheus/pull/17160/commits/925bf092c8ea18a43fe87e518045ed82c812f2b2
. PTAL if you have any free time. Thanks